sklearn/preprocessing/_polynomial.py

```diff
@@ -920,16 +989,57 @@ def transform(self, X):
                 for j in range(degree):
                     mask = X[:, i] < xmin
                     if np.any(mask):
-                        XBS[mask, i * n_splines + j] = (
-                            f_min[j] + (X[mask, i] - xmin) * fp_min[j]
-                        )
+                        linear_extr = f_min[j] + (X[mask, i] - xmin) * fp_min[j]
+                        if use_sparse:
+                            # Note: See comment about SparseEfficiencyWarning above.
+                            XBS_sparse = XBS_sparse.tolil()
+                            XBS_sparse[mask, j] = linear_extr
+                        else:
+                            XBS[mask, i * n_splines + j] = linear_extr
 
                     mask = X[:, i] > xmax
                     if np.any(mask):
                         k = n_splines - 1 - j
-                        XBS[mask, i * n_splines + k] = (
-                            f_max[k] + (X[mask, i] - xmax) * fp_max[k]
-                        )
+                        linear_extr = f_max[k] + (X[mask, i] - xmax) * fp_max[k]
+                        if use_sparse:
+                            # Note: See comment about SparseEfficiencyWarning above.
+                            XBS_sparse = XBS_sparse.tolil()
+                            XBS_sparse[mask, k : k + 1] = linear_extr[:, None]
+                        else:
+                            XBS[mask, i * n_splines + k] = linear_extr
+
+            if use_sparse:
+                if not sparse.isspmatrix_csr(XBS_sparse):
+                    XBS_sparse = XBS_sparse.tocsr()
+                output_list.append(XBS_sparse)
+
+        if use_sparse:
+            # TODO: Remove this conditional error when the minimum supported version of
+            # SciPy is 1.9.2
+            # `scipy.sparse.hstack` breaks in scipy<1.9.2
+            # when `n_features_out_ > max_int32`
+            max_int32 = np.iinfo(np.int32).max
+            all_int32 = True
+            for mat in output_list:
+                all_int32 &= mat.indices.dtype == np.int32
```

Comment:
If we have a CSR matrix of more than `max_int32` no-zeroed lines, then at least an element of `mat.indptr` will not be representable using `np.int32` since we have in this case:
 - `mat.indptr.shape[0] == mat.shape[0] + 1` (true in the general case as well)
 - For all `i` in `[0, mat.indptr.shape[0][`, `mat.indptr[i] < mat.indptr[i+1]` (since we need to store the non-null coordinates in `mat.indices`)

Hence `np.int64` will be used in this case.

Should modify `all_int32` with this condition?
```suggestion
                all_int32 &= mat.indices.dtype == np.int32
                all_int32 &= mat.indptr.dtype == np.int32
```

Reply:
> Isn't one of those sufficient? The sparse matrices are already constructed. We only want to know if at least one of them uses int64.

Yes, that's exact, marking this thread as resolved.